### PR TITLE
SEARCH-2782 commit time as event time

### PR DESCRIPTION
**Pull request by jesty**
This PR is updated because we face an issue when the transaction entry object is not initialized and we cannot retrieve the commit time. This happens when no nodes were added to the transaction because there aren't changes to any node, but we need to produce some events (e.g. permission update).

Before I’m getting the commit time using the nodeDao:
```
Long currentTransactionId = nodeDAO.getCurrentTransactionId(true);
Transaction transaction = nodeDAO.getTxnById(currentTransactionId);
Instant commitTimeMs = Instant.ofEpochMilli(transaction.getCommitTimeMs());
```
The PermissionService does not invoke always NodeDAOImpl.getCurrentTransaction() method, so even if we are inside a transaction the Transaction entry will not be initialized and the method nodeDAO.getCurrentTransactionId() will return null or it will throw an exception, depends on the boolean parameter value. In this case, also the ACL transaction is null. For instance, if you remove an ACL from a node and that ACL type is shared the transaction entry is never initialized.

In this PR I'm purposing a solution checking if the current transaction id is null, and then using the current time.

I prefer the solution above because it avoids exceptions for a trivial reason. Otherwise, we can force the transaction entry initialization invoking nodeDAO.getCurrentTransactionId(true) in PermissionEventGenerator but in this case, we are creating empty transactions and we are relying on a side effect (the getter that inits an object).

### CODE REVIEW

File: repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java [hunk 1]
```diff
@@ -370,13 +381,35 @@ private boolean isFilteredChildAssociation(QName childAssocType, String user)
...
+        return new EventInfo().setTimestamp(getCurrentTransactionTimestamp())
                               .setId(UUID.randomUUID().toString())
                               .setTxnId(AlfrescoTransactionSupport.getTransactionId())
                               .setPrincipal(user)
                               .setSource(URI.create("/" + descriptorService.getCurrentRepositoryDescriptor().getId()));
     }
 
+    private ZonedDateTime getCurrentTransactionTimestamp()
+    {
+        Long currentTransactionId = nodeDAO.getCurrentTransactionId(false);
```
Comment (aborroy) [inline repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java:393]: This method works relying on the execution of this method inside the same Thread that is executing the transaction. If that changes at same point, an alternative method should be adopted.

**Review by aborroy — APPROVED** [review]
LGTM

**Comment by jesty** [comment]
> This PR is updated because we face an issue when the transaction entry object is not initialized and we cannot retrieve the commit time. This happens when no nodes were added to the transaction because there aren't changes to any node, but we need to produce some events (e.g. permission update).
> 
> Before I’m getting the commit time using the nodeDao:
> 
> ```
> Long currentTransactionId = nodeDAO.getCurrentTransactionId(true);
> Transaction transaction = nodeDAO.getTxnById(currentTransactionId);
> Instant commitTimeMs = Instant.ofEpochMilli(transaction.getCommitTimeMs());
> ```
> 
> The PermissionService does not invoke always NodeDAOImpl.getCurrentTransaction() method, so even if we are inside a transaction the Transaction entry will not be initialized and the method nodeDAO.getCurrentTransactionId() will return null or it will throw an exception, depends on the boolean parameter value. In this case, also the ACL transaction is null. For instance, if you remove an ACL from a node and that ACL type is shared the transaction entry is never initialized.
> 
> In this PR I'm purposing a solution checking if the current transaction id is null, and then using the current time.
> 
> I prefer the solution above because it avoids exceptions for a trivial reason. Otherwise, we can force the transaction entry initialization invoking nodeDAO.getCurrentTransactionId(true) in PermissionEventGenerator but in this case, we are creating empty transactions and we are relying on a side effect (the getter that inits an object).

It’s not true that the PermissionService never creates a transaction entry. It creates a new transaction if it is required, I mean, only if there is an actual ACL change to the node. The test was failing because the transaction is null, but the transaction was null because the permission isn’t changed, it is the same before, so we can even avoid producing the event.
IMHO this is an important thing because we can try to make the system lighter avoiding useless events. For instance, two tests have been failed because of:
the test is removing all permissions from a node that hasn’t permission. So the ACL it’s the same.
the test is moving a folder to a new parent with the same ACL as the old parent. So the ACL it’s the same.
In both scenarios, the system is not starting a transaction and relying on that we can avoid producing useless events.

File: repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java (hunk 1 above)
Comment (killerboot) [reply repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java:391]: Afaik the idea is to produce messages from the events (serialise the java object into text and etc) by a separate thread pool of workers. The events are still generated after the transaction commit stage.

**Review by killerboot — APPROVED** [review]

File: repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java [hunk 2]
```diff
@@ -368,13 +379,36 @@ private boolean isFilteredChildAssociation(QName childAssocType, String user)
...
+        ZonedDateTime timestamp;
+        if(currentTransactionId != null)
+        {
+            Transaction transaction = nodeDAO.getTxnById(currentTransactionId);
+            Instant commitTimeMs = Instant.ofEpochMilli(transaction.getCommitTimeMs());
+            timestamp = ZonedDateTime.ofInstant(commitTimeMs, ZoneOffset.UTC);
+        }
+        else 
+        {
+            // Sometimes the transaction entry is not created because no nodes were added to the transaction, but we are producing events anywhere.
```
Comment (killerboot) [inline repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java:401]: If the transaction is not active and nothing was committed to the DB, then there should be no events produced.
I think you can check the transaction state by AlfrescoTransactionSupport.isActualTransactionActive().

**Review by killerboot — CHANGES_REQUESTED** [review]
Can you please check why the events are produced even if the transaction is not committed?
I suspect there is a bug in the event generator or consolidator.

File: repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java (hunk 2 above)
Comment (jesty) [reply repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java:401]: Totally agree, I already did some experiments about it. Should I do it in this PR or in a separate PR/Issue?

Comment (jesty) [reply repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java:401]: Anyway, isActualTransactionActive doesn't work in this case because the transaction it is active, but it is never used so AlfrescoTransactionSupport.getResource(KEY_TRANSACTION) is null.

**Comment by jesty** [comment]
> Can you please check why the events are produced even if the transaction is not committed?
> I suspect there is a bug in the event generator or consolidator.

The transaction is committed, but the TransactionEntity is never created and stored in the database because there isn't a real change. For instance, if I remove all permission from a node that hasn't permission, or if I change permission to a node setting the same ACL it has before the change, the transaction is created and committed, but TransactionEntity will never create and stored in the database. This happens because TransactionEntity will be initialized only when the system adds something to a transaction and if there aren't real updates, the system doesn't add anything to the transaction.

**Comment by killerboot** [comment]
> > Can you please check why the events are produced even if the transaction is not committed?
> > I suspect there is a bug in the event generator or consolidator.
> 
> The transaction is committed, but the TransactionEntity is never created and stored in the database because there isn't a real change. For instance, if I remove all permission from a node that hasn't permission, or if I change permission to a node setting the same ACL it has before the change, the transaction is created and committed, but TransactionEntity will never create and stored in the database. This happens because TransactionEntity will be initialized only when the system adds something to a transaction and if there aren't real updates, the system doesn't add anything to the transaction.

In this case the event consolidator should not produce any events as there was no change to nodes or permissions.

**Comment by jesty** [comment]
> > > Can you please check why the events are produced even if the transaction is not committed?
> > > I suspect there is a bug in the event generator or consolidator.
> > 
> > 
> > The transaction is committed, but the TransactionEntity is never created and stored in the database because there isn't a real change. For instance, if I remove all permission from a node that hasn't permission, or if I change permission to a node setting the same ACL it has before the change, the transaction is created and committed, but TransactionEntity will never create and stored in the database. This happens because TransactionEntity will be initialized only when the system adds something to a transaction and if there aren't real updates, the system doesn't add anything to the transaction.
> 
> In this case the event consolidator should not produce any events as there was no change to nodes or permissions.

So, I identified 4 scenarios:

- Node Transaction
- ACL changeset (Enterprise only)
- Node Transaction and ACL Changeset (Enterprise only)
- No transaction, in my test this happens only in few corner cases, I think no so common, for instance when you remove permissions from a node that hasn't any permission.

I updated this PR by adding a method to retrieve the ACL changeset object, this method will be used in the enterprise version to retrieve the ACL Changeset commit time.

File: repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAO.java [hunk 3]
```diff
@@ -1,28 +1,28 @@
...
+ * 
+ * Alfresco is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ * 
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Alfresco. If not, see <http://www.gnu.org/licenses/>.
+ * #L%
+ */
```
Comment (aborroy) [inline repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAO.java:25]: What is the change in these lines?

File: repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAOImpl.java [hunk 4]
```diff
@@ -1,28 +1,28 @@
...
+ * 
+ * Alfresco is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ * 
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Alfresco. If not, see <http://www.gnu.org/licenses/>.
+ * #L%
+ */
```
Comment (aborroy) [inline repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAOImpl.java:25]: Again, it's confusing adding this kind of changes to a commit...

File: repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAOImpl.java [hunk 5]
```diff
@@ -1662,9 +1663,29 @@ public void beforeCommit(boolean readOnly)
...
             // Update it
             long commitTimeMs = System.currentTimeMillis();
+            AlfrescoTransactionSupport.bindResource(RESOURCE_KEY_ACL_CHANGE_SET_COMMIT_TIME_MS, commitTimeMs);
             aclCrudDAO.updateAclChangeSet(changeSetId, commitTimeMs);
         }
     }
+
+    /**
+     * Retrieve the value from the current thread local, avoiding the access to the database.
+     * @return
```
Comment (aborroy) [inline repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAOImpl.java:1673]: What happens if this method returns null?

**Review by aborroy — APPROVED** [review]
Minor comments

File: repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAO.java [hunk 6]
```diff
@@ -182,4 +182,9 @@
      * @return Long
      */
     public Long getMaxChangeSetIdByCommitTime(long maxCommitTime);
+
+    /**
+     * @return the current AclChangeSet created for the current ACL change
```
Comment (killerboot) [inline repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAO.java:187]: This javadoc is not clear. "Current" means for current transaction, if it is present. The return value can be null.

The AclChangeSet is never used in this interface, do you know why?

**Review by killerboot — CHANGES_REQUESTED** [review]
The events should **not** be generated for empty transactions. Please fix that.

File: repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAO.java (hunk 6 above)
Comment (jesty) [reply repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAO.java:187]: I can change to: "@return the current AclChangeSet created for the current transaction. It can be null if a transaction is not active."
AclChangeSet is used only inside the implementation. In the interface it uses only field of this object, for instance in getMaxChangeSetIdByCommitTime and getMaxChangeSetCommitTime.

File: repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAO.java (hunk 3 above)
Comment (jesty) [reply repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAO.java:25]: sorry, it is a formatting change, I'll remove both

File: repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAO.java (hunk 6 above)
Comment (killerboot) [reply repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAO.java:187]: AclChangeSet is a DAO layer entity and should not be exposed via the interface :)

File: repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAOImpl.java (hunk 4 above)
Comment (jesty) [reply repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAOImpl.java:25]: sorry, it is a formatting change and IntelliJ doesn't show to me, I'll remove both

File: repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAOImpl.java (hunk 5 above)
Comment (jesty) [reply repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAOImpl.java:1673]: I'll change the Javadoc in the interface, if null it means that there isn't a transaction active.

File: repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAO.java (hunk 6 above)
Comment (jesty) [reply repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAO.java]: I apologize because they are in the same package the layering wasn't clear to me. Anyway I did a refactoring in order to have a method like  getMaxChangeSetIdByCommitTime

File: repository/src/main/java/org/alfresco/repo/domain/node/AbstractNodeDAOImpl.java [hunk 7]
```diff
@@ -604,6 +605,17 @@ private TransactionEntity getCurrentTransaction()
         return txn;
     }
     
+    public Long getCurrentTransactionCommitTime()
```
Comment (jesty) [inline repository/src/main/java/org/alfresco/repo/domain/node/AbstractNodeDAOImpl.java:608]: In order to avoid useless access to the database, I decide to use ThreadLocal in order to retrieve the current commit time, as did for the id, instead of reading it from the database.

File: repository/src/main/java/org/alfresco/repo/domain/node/AbstractNodeDAOImpl.java [hunk 8]
```diff
@@ -563,6 +563,7 @@ public void beforeCommit(boolean readOnly)
             Long txnId = txn.getId();
             // Update it
             Long now = System.currentTimeMillis();
+            txn.setCommitTimeMs(now);
```
Comment (killerboot) [inline repository/src/main/java/org/alfresco/repo/domain/node/AbstractNodeDAOImpl.java:566]: I think this is done in updateTransaction method.

File: repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAO.java [hunk 9]
```diff
@@ -182,4 +182,11 @@
      * @return Long
      */
     public Long getMaxChangeSetIdByCommitTime(long maxCommitTime);
+
+    /**
+    * @return                 the commit time of the current ACL change set entry or <tt>null</tt> if
+    *                         there have not been any modifications.
+    */ 
+    Long getCurrentChangeSetCommitTime();
```
Comment (killerboot) [inline repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAO.java:190]: The method should be "public", right?

File: repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAOImpl.java [hunk 10]
```diff
@@ -1636,7 +1636,8 @@ public Acl createLayeredAcl(Long indirectedAcl)
         return changes;
     }
 
-    private static final String RESOURCE_KEY_ACL_CHANGE_SET_ID = "acl.change.set.id";
+    public static final String RESOURCE_KEY_ACL_CHANGE_SET_ID = "acl.change.set.id";
```
Comment (killerboot) [inline repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAOImpl.java:1639]: Why did you make them public?

File: repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java [hunk 11]
```diff
@@ -392,36 +425,41 @@ protected void onShutdown(ApplicationEvent applicationEvent)
...
+                    // Node events
+                    for (Map.Entry<NodeRef, EventConsolidator> entry : consolidators.getNodes().entrySet())
+                    {
+                        EventConsolidator eventConsolidator = entry.getValue();
+                        sendEvent(entry.getKey(), eventConsolidator);
+                    }
+
+                    // Child assoc events
+                    for (Map.Entry<ChildAssociationRef, ChildAssociationEventConsolidator> entry : consolidators.getChildAssocs()
+                                                                                                           .entrySet())
```
Comment (killerboot) [inline repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java:443]: This formatting looks very odd, please move everything in one line.

File: repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java [hunk 12]
```diff
@@ -392,36 +425,41 @@ protected void onShutdown(ApplicationEvent applicationEvent)
...
+                    {
+                        ChildAssociationEventConsolidator eventConsolidator = entry.getValue();
+                        sendEvent(entry.getKey(), eventConsolidator);
+                    }
+
+                    // Peer assoc events
+                    for (Map.Entry<AssociationRef, PeerAssociationEventConsolidator> entry : consolidators.getPeerAssocs()
+                                                                                                     .entrySet())
```
Comment (killerboot) [inline repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java:451]: This formatting looks very odd, please move everything in one line.

File: repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java (hunk 2 above)
Comment (killerboot) [reply repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java:400]: Is it correct that after the changes in EventTransactionListener in this class this code will never be executed? If so, can this if-else block be removed?

**Review by killerboot — CHANGES_REQUESTED** [review]

File: repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAO.java (hunk 9 above)
Comment (jesty) [reply repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAO.java:190]: in interfaces, the default modifier is public, so as IntelliJ uses to say: "Modified ‘public’ is redundant for interface methods" :)

File: repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java (hunk 2 above)
Comment (jesty) [reply repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java:400]: removed

File: repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAOImpl.java (hunk 10 above)
Comment (jesty) [reply repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAOImpl.java]: I used RESOURCE_KEY_ACL_CHANGE_SET_COMMIT_TIME_MS in a test in the enterprise repository. In order to don't have similar constants with different modifiers, I used the same. Anyway, especially because they are in two different repositories, I prefer to avoid the coupling in the test, so I replaced the public with private and I'll use the string in the test.

File: repository/src/main/java/org/alfresco/repo/domain/node/AbstractNodeDAOImpl.java (hunk 8 above)
Comment (jesty) [reply repository/src/main/java/org/alfresco/repo/domain/node/AbstractNodeDAOImpl.java:566]: not exactly, updateTransaction method creates a new object and uses it for the update, while I need to update the object in memory.
```
  @Override
    protected int updateTransaction(Long txnId, Long commitTimeMs)
    {
        TransactionEntity transaction = new TransactionEntity();
        transaction.setId(txnId);
        transaction.setCommitTimeMs(commitTimeMs);
        return template.update(UPDATE_TRANSACTION_COMMIT_TIME, transaction);
    }
```

File: repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java (hunk 12 above)
Comment (jesty) [reply repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java]: done

File: repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAO.java (hunk 9 above)
Comment (killerboot) [reply repository/src/main/java/org/alfresco/repo/domain/permissions/AclDAO.java:190]: Please make it consistent with the rest of the interface.

File: repository/src/main/java/org/alfresco/repo/domain/node/NodeDAO.java [hunk 13]
```diff
@@ -75,6 +75,13 @@
     /*
      * Transaction
      */
+
+    /**
+     * @return                 the commit time of the current transaction entry or <tt>null</tt> if
+     *                         there have not been any modifications to nodes registered in the
+     *                         transaction.
+     */
+    Long getCurrentTransactionCommitTime();
```
Comment (killerboot) [inline repository/src/main/java/org/alfresco/repo/domain/node/NodeDAO.java:84]: Please make the method visibility consistent with the rest of the interface.

**Review by killerboot — APPROVED** [review]
